examples/bulk_base_api.cxx

```diff
@@ -0,0 +1,338 @@
+        }
+        if (const auto* val = getenv("CB_PASSWORD"); val != nullptr && val[0] != '\0') {
+            arguments.password = val;
+        }
+        if (const auto* val = getenv("CB_BUCKET_NAME"); val != nullptr && val[0] != '\0') {
+            arguments.bucket_name = val;
+        }
+        if (const auto* val = getenv("CB_SCOPE_NAME"); val != nullptr && val[0] != '\0') {
+            arguments.scope_name = val;
+        }
+        if (const auto* val = getenv("CB_COLLECTION_NAME"); val != nullptr && val[0] != '\0') {
+            arguments.collection_name = val;
+        }
+        if (const auto* val = getenv("CB_NUMBER_OF_OPERATIONS"); val != nullptr && val[0] != '\0') {
+            char* end = nullptr;
+            auto int_val = std::strtoul(val, &end, 10);
+            if (end != val) {
+                arguments.number_of_operations = int_val;
+            }
+        }
+        if (const auto* val = getenv("CB_DOCUMENT_BODY_SIZE"); val != nullptr && val[0] != '\0') {
+            char* end = nullptr;
+            auto int_val = std::strtoul(val, &end, 10);
+            if (end != val) {
+                arguments.document_body_size = int_val;
+            }
+        }
+        return arguments;
+    }
+};
+
+std::string
+random_text(std::size_t length)
+{
+    std::string alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+    static thread_local std::mt19937_64 gen{ std::random_device()() };
+    std::uniform_int_distribution<std::size_t> dis(0, alphabet.size() - 1);
+    std::string text(length, '-');
+    for (std::size_t i = 0; i < length; ++i) {
+        text[i] = alphabet[dis(gen)];
+    }
+    return text;
+}
+
+auto
+generate_document(std::size_t document_body_size) -> tao::json::value
+{
+    return {
+        { "size", document_body_size },
+        { "text", random_text(document_body_size) },
+    };
+}
+
+void
+run_workload_sequential(const couchbase::collection& collection, const program_arguments& arguments)
+{
+    fmt::print("\n===== SEQUENTIAL\n");
+
+    const std::string document_id_prefix{ "sequential" };
+    std::vector<std::string> document_ids;
+    document_ids.reserve(arguments.number_of_operations);
+    for (std::size_t i = 0; i < arguments.number_of_operations; ++i) {
+        document_ids.emplace_back(fmt::format("{}_{:06d}", document_id_prefix, i));
+    }
+    fmt::print("Using {} IDs in interval [\"{}\"...\"{}\"]\n", document_ids.size(), document_ids[0], document_ids[document_ids.size() - 1]);
+
+    const auto document = generate_document(arguments.document_body_size);
+
+    auto start = std::chrono::system_clock::now();
+
+    {
+        std::map<std::string, std::size_t> errors;
+
+        auto exec_start = std::chrono::system_clock::now();
+        for (std::size_t i = 0; i < arguments.number_of_operations; ++i) {
+            auto [ctx, result] = collection.upsert(document_ids[i], document, {}).get();
+            errors[ctx.ec().message()]++;
```

Comment:
not really, I also want to report success code here too, just to make sure it matches with the expected number of operations.

Reply:
Fair enough.  It holds that even if users see `Undefined error: 0: 1000` instead of `success: 1000`(like I do on my local machine) that they will still have all the details on how to do bulk ops. 👍